### ISSUE TITLE
How do labels work for multiple choice question Checkbox?

### ISSUE THREAD

**Issue by shinuito** [maths/moodle-qtype_stack#639]
Hi,

I've been attempting to use the multiple choice question type Checkbox and following the information in https://github.com/maths/moodle-qtype_stack/blob/master/doc/en/Authoring/Multiple_choice_questions.md. For version info "Your site is running STACK version 2020070100 (descriptive, distrib, simplex, stats)."

I'm having a lot of trouble using the labels feature to display the options, specifically with latex/mathjax. If I just display the values, they end up unreasonably close to the checkboxes while using inline display, while with display equation they are very far away and centering makes it a bit confusing which box they are aligned with.

I would like to use the labels to 
- (a) render latex expressions which insert maxima expressions (random variables). In particular say, with intervals, either to use stacks oo,oc,co,cc and inf or to use \infty and have it rendered properly.
- (b) hopefully control the spacing so expression have at least some empty space between them and the checkbox.

The documentation isn't very clear on this. It kind of suggests that the whole point is to be able to display maxima/stack expressions as options "Question authors should consider using the Moodle MCQ question type in addition to these facilities for purely text-based answers." 

However, how do I display expressions while adding spacing?

Suppose <code>ta:[[A,true, "(bb/cc,inf)"],[B,false,"(-inf -bb/cc]" ]]</code> as say a small example. Strings or use of cc,co, etc are necessary not to confuse the bracketing as parts of lists. However with the strings it wont render the variables. 

Can someone give a working example with latex labels which also use a random variable which is evaluated?

An example in the documentation using latex seems to be brute forcing the inline equation \\ ( and \\ )  into the label rather than using the options.

**Comment by shinuito** [comment]
I might have an answer, but it seems like the option LaTeXinline doesn't really work on STACK version 2020070100.

Suppose I want to display a label where the object is an interval \((-\infty,a]\) where a is a random variable, then I can brute force in the latex inline into the label and the value of a by using <code>sconcat</code> and <code>string</code>. It seems like concat is banned.

So one entry of ta might look like

<pre><code>[oc(-inf,a),true,sconcat("\\(\\, -\infty, ",string(a),"]\\)")] </code></pre>

**Comment by sangwinc** [comment]
There is a function `stack_disp`
https://github.com/maths/moodle-qtype_stack/blob/master/stack/maxima/stackmaxima.mac#L571
I didn't originally intend this to be used by teachers, and it has some problems.  E.g. in this case

`[oc(-inf,a),true,stack_disp(oc(-inf,a),"i")]` 

won't display properly.  The reason is that mismatching brackets are a problem for the PHP-Maxima connection, and there are times when we have to protect them.  So, we actually return 

`\([oc(-inf,a),true,"!LEFTR! -\\infty ,\\, a!RIGHTSQ!"]\)`

Within CASText these are "sorted out", but having just created a checkbox question they are not, because the conversion of values into displayed strings  

The conversions are defined here: 
https://github.com/maths/moodle-qtype_stack/blob/master/locallib.php#L58

The reason this conversion is not applied here:
https://github.com/maths/moodle-qtype_stack/blob/master/stack/input/dropdown/dropdown.class.php#L172
is that we expected the optional argument to be any kind of general "string" and not the LaTeX generated by Maxima for STACK! 

I'll review if adding these conversions will break any other strings, if not we might as well support this kind of use.  

I should just note in the example you use you might as well use 

`[oc(-inf,a),true)]` 

since the string part is optional.  Of course if you wanted more text, e.g.

`[oc(-inf,a),true,sconcat("The half interval: ", stack_disp(oc(-inf,a),"i"))]` 

Then we still need this conversion.  Thanks for raising this.  I know people are using MCQ with some LaTeX in it.

Should we convert the format of these optional strings to full CAStext?  Probably would be useful but it is going to be very difficult to make sure all the information (the question variable values) are available to the input method. That probably needs to wait for "input2" developments.

(Just as a reminder for anyone finding this issue in the future, you can't use LaTeX in the dropdowns currently.  Some unicode alternatives are defined in the docs....)

**Comment by shinuito** [comment]
@sangwinc Thanks for the quick reply and considering it.

I guess my issue initially with just using <code>[oc(-inf,a),true)]</code> and not the optional label, is that I don't really have any control over the display, just LaTeXinline or LaTeXdisplay. The inline was putting the options really close to the checkbox (it's the option I'm going with),
![image](https://user-images.githubusercontent.com/26803867/95070845-9f5d0500-0700-11eb-8311-a7e6a014ca8b.png)
whereas the display equation, everything is very far away from the check box, and not aligned with it
![image](https://user-images.githubusercontent.com/26803867/95070936-cd424980-0700-11eb-9c91-7c153d3da510.png)

I will have to test out what you're saying above about <code>stack_disp</code>. I'm guessing then the 'i' is for inline.

Maybe even a simpler issue I'm having with displaying random variables in the strings for instance is if the variable is something like <code>b:-7/5</code>. Then trying to insert it into the label with just <code>string(b)</code> puts in extra brackets -(7/5). The <code>stack_disp(b,"i")</code> does actually display it nicely as a fraction, but I can't concatenate it I suppose inside another LaTeX environment.

Actually the best I've gotten working is probably a combination of the above ideas.
![image](https://user-images.githubusercontent.com/26803867/95072702-86098800-0703-11eb-9be5-82db73ae266b.png)

I used the following for now as a label <pre><code>sconcat("\\(\\,(-\\infty,\\)",stack_disp(b,"i"),"\\(]\\)")</code></pre>

In theory I could substitute <pre><code>\\,</pre></code> for <pre><code>\\quad</pre></code> or something else (I need to figure out what moodle is allowing) to edit the spacing.

**Comment by sangwinc** [comment]
I agree, we really do need some space there. I'll add a little bit of space for the default as well.

**Comment by aharjula** [comment]
As to that space, sure that is something that should be added, and the CSS logic for adding that should not be a problem.

As to `stack_display(b,"i")` try `stack_display(b,"")` and it does not add the wrappers and is usable within a latex environment. The second argument for that function just controls the type of wrappers with these values `"", "i", "d", "id"`...

While `stack_disp()` is the correct way, no one says that one could not use `tex1()` to generate the LaTeX fragment in these particular cases. `tex1` does not protect those brackets, and I don't really think we need to protect them anymore as the 4.3 update changed the Maxima-PHP to something that shoudl nto have any issues with any brackets.

Overall my opinnion is that if and when we replace the input system with something new the labels should allow direct CASText with all of its capabilities. The current input2-system prototype includes two types of MCQ-inputs one that eats this old form of option definition and another that receives the options as a separate structure though the input options and that structure assumes that the label is CASText and renders it as such.

Finally, I really hope that beyond CSS-style tuning nothing else gets done to the label rendering logic of the old system especially not during COVID as it is highly unlikely that we could automatically fix any old materials that have hacked labels together usign various means and could thus break materials. I would really not want to see anything messing up any more materials in the current situation as people already have their hands full with the increased interest.

**Comment by shinuito** [comment]
>Finally, I really hope that beyond CSS-style tuning nothing else gets done to the label rendering logic of the old system especially not during COVID as it is highly unlikely that we could automatically fix any old materials that have hacked labels together usign various means and could thus break materials. I would really not want to see anything messing up any more materials in the current situation as people already have their hands full with the increased interest.

That's completely reasonable. 

Thanks for the feedback regarding the use of `stack_display` and `tex1`. I'll play around a bit more with it. For some reason, even though I recognised "i" option as being for inline, I wasn't connecting it to the function inserting the inline brackets!
